### ISSUE TITLE
Include clang-query into clangd

### ISSUE THREAD

**Issue by Febbe** [clangd/clangd#1945]
This is a proposal which I already submitted to the [llvm/llvm-project](https://github.com/llvm/llvm-project), but it looks like, that none is watching the issues for clangd there.

I would like to search the AST via the AST Matcher DSL from clang from the search bar of any editor, utilizing clangd.

My desired workflow would be, 
 - to select AST Matcher Search
 - to select code
 - Press Ctrl + Shift + F / Ctrl + Shift + H
 - Modify the generated ASTMatcher sequence
 - Press Enter
 - All matching AST sequences are ahown
 - Optionally replace them with a sub AST

**Comment by Febbe** [comment]
@sam-mccall I have interest, to work on this myself, is there a chance this proposal will be accepted?
Otherwise, I think about creating a standalone version, with a vscode plugin.

    Editor support: How many users will the feature be available to?
     - It should be a search engine, so every editor, which allows customization of searches, should be able to use/implement it.
    Standardization: Is the feature stable? Is it likely to be adopted by more editors over time?
     - Search engines are, all Editors support them.
    Utility: Does the feature provide a lot of value?
     - Yes, developers often find a bug or code snippets which aren't ideal, and they want to refactor them. This would add a convenient way to do this automatically. It also could be used for many more refactoring methods, like changing the header of a function.
    Complexity: Is this hard to implement in clangd, or constrain future work? Is the protocol complicated?
     - Most of the work already exist in clang-query, this only requires adding its (e)DSL interpreter to clangd.
     - A protocol must be defined, but it will most likely be an extension of the show AST protocol
     - We have to implement an implementation for the vscode plugin. Either by extending the search engine, when this is possible, or by adding a new widget for it.
     
Overall, the benefits are in my eyes larger than the additional maintenance burden.
What do you think about that?

**Comment by HighCommander4** [comment]
I don't fully understand this feature based on the description given. What is the "generated ASTMatcher sequence"?

Could you give a concrete example with code?

**Comment by HighCommander4** [comment]
Thank you for elaborating. This sounds like a really neat feature.

I would suggest starting with a simpler building block: an AST Matcher search feature where you type the query string manually, and clangd returns matches in the open file.

Then, other parts of the feature can be layered on top.

One idea for a client-side enhancement that might be really neat would be to have the client provide a graphical UI for visualizing and editing query strings, that shows the tree structure of the query and stuff like that.

For the part of the feature that involves turning an AST subtree into an AST matcher query that matches that subtree -- if I'm not mistaken, that's not functionality that exists today, it needs to be implemented, right? It sounds somewhat non-trivial to me. For that, it's probably best to start by writing an RFC thread in the [clang frontend forum](https://discourse.llvm.org/c/clang/6), proposing to add this capability to the AST Matchers library. Then once it's in place we can talk about clangd making use of it. (Work on this can be done in parallel with work on the clangd feature to search for a user-provided query string.)

**Comment by HighCommander4** [comment]
> @sam-mccall I have interest, to work on this myself, is there a chance this proposal will be accepted?

I obviously can't speak for Sam, but I'm personally supportive of this feature, especially if it's developed and added in pieces as described above.

**Comment by torshepherd** [comment]
Hi all, I have a lot of interest in working on this as well. A lot of my current work involves large-scale searching for patterns and refactoring, and clang-tidies are simply too slow of a loop.

I think it makes sense for it to be integrated into clangd, but it doesn't really need to depend on much machinery in clangd (the LSP is useful as a communication layer is all).

I would propose that we form some sort of implementation group on the LLVM discord and start planning there.

**Comment by torshepherd** [comment]
I'm taking a crack at your initial approach @HighCommander4 , but clang-query is meant to work with `ASTUnit`s, but clangd uses `ParsedAST`. I'll need to reconcile them somehow.

Otherwise this first step to integration should be very straightforward - just adds a new extension point callback for `matchAst`, parse a query matcher from string (provided by clang-query already), run the matcher, and return a vector of ASTNodes, serialize to JSON, and respond from the LSP

**Comment by HighCommander4** [comment]
> clang-query is meant to work with `ASTUnit`s, but clangd uses `ParsedAST`. I'll need to reconcile them somehow.

`ASTUnit` looks like a fairly thin wrapper around `ASTContext` and a few related pieces of state (which is basically what `ParsedAST` is as well).

Operations on `ASTUnit` should translate fairly easily to operations on `ParsedAST` (most obviously, getting at the `ASTContext` with `getASTContext()`, but other things should have straightforward equivalents as well, e.g. for `ASTUnit::getDiagnostics()`, `ParsedAST` can expose `Clang->getDiagnostics()` if needed).

**Comment by Febbe** [comment]
I already started, but hadn't time to further work on this, just pushed my changes to `https://github.com/Febbe/llvm-project/tree/clangd_ast_search`

**Comment by torshepherd** [comment]
Took a look, this is basically how I was gonna approach it too. I'll draft up a quick version of the client to match your server endpoint implementation.

The problem is that vscode unfortunately doesn't have a great way to extend the "find in file" UI. So the only way to do this is with a sidebar tree view such as the Dump AST does now

**Comment by Febbe** [comment]
> The problem is that vscode unfortunately doesn't have a great way to extend the "find in file" UI. So the only way to do this is with a sidebar tree view such as the Dump AST does now

That feature is already on the way 😁: https://github.com/microsoft/vscode/issues/59921

**Comment by torshepherd** [comment]
@Febbe since the issue was added to MSFT VSCode backlog, seems like it will eventually be available. Do you have time to pick back up your branch this year?

**Comment by Febbe** [comment]
@torshepherd I finally found some time to work on this. My code now compiles, but I haven't tested it yet.

The first stage "type an ast matcher sequence and search for it" should be implemented if it works right ahead.

For now only one traversal kind and one single matcherID ("root") is supported.

The search result is returned in an array containing the results. Each result is currently also an array, containing a ASTNodes for each MatcherID. 
(Next step is, to make the Result a pair of matcherId and AstNode)
